test/headers_test.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2020 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package test
+
+import (
+	"reflect"
+	"testing"
+	"time"
+
+	natsserver "github.com/nats-io/nats-server/v2/test"
+	"github.com/nats-io/nats.go"
+)
+
+func TestBasicHeaders(t *testing.T) {
+	s := RunServerOnPort(-1)
+	defer s.Shutdown()
+
+	nc, err := nats.Connect(s.ClientURL())
+	if err != nil {
+		t.Fatalf("Error connecting to server: %v", err)
+	}
+	defer nc.Close()
+
+	subject := "headers.test"
+	sub, err := nc.SubscribeSync(subject)
+	if err != nil {
+		t.Fatalf("Could not subscribe to %q: %v", subject, err)
+	}
+	defer sub.Unsubscribe()
+
+	m := nats.NewMsg(subject)
+	m.Header.Add("Accept-Encoding", "json")
+	m.Header.Add("Authorization", "s3cr3t")
+	m.Data = []byte("Hello Headers!")
```

Comment:
not a huge fan of accessing these struct keys directly, but I guess that's what we do already so more of the same is ok.  Would have been really nice if nats.Msg could be a interface

Reply:
Use this directly - https://golang.org/pkg/net/http/#Header

Reply:
Not sure how that relates, my point is I wish the Msg being sent around was a interface and not a struct, it would make it much easier to support things like different encodings etc.

But not something we can realistically fix, was just a observation

Reply:
We are already working on a v2 generic client that is more focused on streams and services and will be JetStream aware, so that would be the place to do so.

Not following the different encodings though.

Reply:
> We are already working on a v2 generic client that is more focused on streams and services and will be JetStream aware, so that would be the place to do so.

News to me :)

Reply:
Just has a brainstorming meeting and some code by Wally and I a couple of weeks ago. Still very early, so plenty of time for review and input.